keps/sig-auth/2843-request-ctx-in-token-review/README.md

```diff
@@ -0,0 +1,109 @@
+# KEP-2843: Forward Additional Request Metadata in TokenReviews
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+<!-- /toc -->
+
+## Summary
+
+This KEP proposes enhancing the TokenReview API to convey additional request
+context to the backing Identity Provider (IDP).
+
+## Motivation
+
+Today, certain request attributes that are known to kube-apiserver may be
+relevant to the backend of a TokenReview webhook. For example:
+
+* Was the request made over https?
+* What are the IP addresses of the peer and origin of the request?
+
+These attributes can be used to enhance audit logs in the IDP, determine access
+levels (such as whether the request originates from a trusted network), or
+enforce security policy centrally (such as disallowing bearer token
+authentication over insecure protocols).
+
+### Goals
+
+* Forward additional request metadata to TokenReview. 
+
+### Non-Goals
+
+* Forward resource metadata to TokenReview: The object of the request (e.g.
+  which verb, which namespace, which pod) should remain unknown to the IDP.
+* Forward addition request metadata to SubjectAccessReview or AdmissionReview:
+  That's what User.ExtraInfo is for.
+
+## Proposal
+
+We will extend TokenReviewSpec to add an additional field:
+
+```golang
+
+type TokenReviewSpec struct {
+  ...
+
+  // RequestAttributes is a collection of request attributes that may be relevant
+  // to a TokenReview backend. These attributes should generally be stable over
+  // multiple requests.
+  //
+  // Note: TokenReview webhook backends should handle abscences of this field or
+  // subfields of the RequestAttribute object. This data may be abscent when the
+  // API Server initiating the TokenReview is old or the data was unavailable.
+  RequestAttributes RequestAttributes
+}
+...
+
+type RequestAttributes struct {  
+  // The HTTP request `Host` header value.
+  Host string
+
+  // The HTTP URL scheme, such as `http` and `https`.
+  Scheme string
+
+  // The IP address of the remote peer.
+  //
+  // Note: This is only the IP address of the originator of the request, when
+  // the network path of the request does not traverse any hops that perform
+  // SNAT. For example, if the request path traverses a simple proxy, this
+  // will be the address of the proxy.
+  PeerIP string
+
+  // Additional standard headers used by frontproxies to forwarded request
+  // attributes to backends. Currently, the only forwarded headers are:
+  // * X-Forwarded-For
+  // * X-Forwarded-Proto
+  ProxyHeaders map[string][]string
```

Comment:
Added since we support this in audit logs. In the naive implementation, this field is user controlled when no L7 proxies are used. Some options:
* We can forward this to the backend and let the backend determine what combination of ProxyHeaders and fields to trust.
* Somehow determine that proxy headers should be trusted, and then forward.